HLTrigger/JetMET/plugins/HLTScoutingPFProducer.cc

```diff
@@ -0,0 +1,249 @@
+    Handle<reco::PFCandidateCollection> pfCandidateCollection;
+    if(doCandidates && !iEvent.getByToken(pfCandidateCollection_, pfCandidateCollection)){
+        edm::LogError ("HLTScoutingPFProducer") << "invalid collection: pfCandidateCollection" << "\n";
+        return;
+    }
+    //get vertices
+    Handle<reco::VertexCollection> vertexCollection;
+    if(!iEvent.getByToken(vertexCollection_, vertexCollection)){
+        edm::LogError ("HLTScoutingPFProducer") << "invalid collection: vertexCollection" << "\n";
+        return;
+    }
+    //get rho
+    Handle<double> rho;
+    if(!iEvent.getByToken(rho_, rho)){
+        edm::LogError ("HLTScoutingPFProducer") << "invalid collection: rho" << "\n";
+        return;
+    }
+    std::auto_ptr<double> outRho(new double(*rho));
+    //get MET
+    Handle<reco::METCollection> metCollection;
+    if(doMet && !iEvent.getByToken(metCollection_, metCollection)){
+        edm::LogError ("HLTScoutingPFProducer") << "invalid collection: metCollection" << "\n";
+        return;
+    }
+
+    //produce vertices
+    std::auto_ptr<ScoutingVertexCollection> outVertices(new ScoutingVertexCollection());
+    for(auto &vtx : *vertexCollection){
+        outVertices->emplace_back(
+                vtx.x(), vtx.y(), vtx.z(), vtx.zError()
+                );
+    }
+
+    //produce PF candidates
+    std::auto_ptr<ScoutingParticleCollection> outPFCandidates(new ScoutingParticleCollection());
+    if(doCandidates){
+        for(auto &cand : *pfCandidateCollection){
+            if(cand.pt() > pfCandidatePtCut){
+                outPFCandidates->emplace_back(
+                        cand.pt(), cand.eta(), cand.phi(), cand.mass(), cand.pdgId()
+                        );
+            }
+        }
+    }
+    
+    //produce PF jets
+    std::auto_ptr<ScoutingPFJetCollection> outPFJets(new ScoutingPFJetCollection());
+    for(auto &jet : *pfJetCollection){
+        if(jet.pt() < pfJetPtCut || fabs(jet.eta()) > pfJetEtaCut) continue;
+        //find the jet tag corresponding to the jet
+        float tagValue = -20;
+        float minDR = 0.1;
+        if(doJetTags){
+            for(auto &tag : *pfJetTagCollection){
+                float dR = reco::deltaR(jet, *(tag.first));
+                if(dR < minDR){
+                    minDR = dR;
+                    tagValue = tag.second;
+                }
+            }
+        }
+        //get the PF constituents of the jet
+        std::vector<int> candIndices;
+        if(doCandidates){
+            for(auto &cand : jet.getPFConstituents()){
+                if(cand->pt() > pfCandidatePtCut){
+                    //search for the candidate in the collection
+                    float minDR = 0.01;
+                    int matchIndex = -1;
+                    int outIndex = 0;
+                    for(auto &outCand : *outPFCandidates){
+                        float dR = sqrt(pow(cand->eta() - outCand.eta(), 2) + pow(cand->phi() - outCand.phi(), 2));
```

Comment:
trivial optimisation: instead of computing the square root for every check, you should compute 
```
const float minDR2 = minDR * minDR;
```
once, and compare it with
```
float dR2 = pow(cand->eta() - outCand.eta(), 2) + pow(cand->phi() - outCand.phi(), 2);
```